### ISSUE TITLE
Add automatic tuning flags to utils.data.dataloader

### ISSUE THREAD

**Issue by OBannon37** [pytorch/pytorch#23048]
## 🚀 Feature
Adding automatic tuning flags for batch_size and num_workers to torch.utils.data.dataloader

## Motivation

We'd like to help users to get good performance (throughput) in their data loading jobs. This will allow users to focus on defining the logic for the model inference.

Today, users have to tune batch_size and num_workers manually when they hit performance issues. Tuning requires some expert knowledge of PyTorch and systems design. Even with this knowledge, users may spend hours or days benchmarking, profiling, identifying bottlenecks, and tweaking parameters to improve performance.

This feature is especially relevant to IO bound pipelines, commonly found in inference workloads, where images are not cached in memory and must be loaded from disk. Optimal batch_size and num_workers configs can improve an imgs/sec metric by as much as 10x.

## Pitch

Just as Tensorflow has introduced the tf.data.experimental.AUTOTUNE feature for batch_size, prefetch_size, and num_parallel_calls in datasets, PyTorch should include a similar feature for torch.utils.data.dataloader.

## Alternatives

If autotuning flags are not introduced, users can guess at the optimal configs, or hyperparameter tune these configs before running full training / inference jobs on big data.

In practice, setting batch_size to be the largest possible without running into OOM errors works well, as does setting num_workers equal to the number of cores on the machine. However, these configs are not always optimal, and may not take advantage of features like hyperthreading.

Hyperparameter tuning can also work to discover optimal parameters, but lacks information about buffer sizes and has to deal with more noise and variance in imgs/sec result metrics. Additionally, extensive engineering is required to enable online tuning; otherwise, valid inference results may be discarded. 

## Additional context

My Questions
* Are there any features like this in the PyTorch Pipeline?
* Would this be useful to other folks in the contributor or user community of PyTorch?
* Has anyone thought about autotuning before and/or have ideas/stubs of implementation?

**Comment by cpuhrsch** [comment]
Hey @OBannon37 - automatic tuning is a wide-ranging topic. If we were to introduce a concept like that it'd be shared across various components of the platform and we'd have to determine the point of highest leverage. The DataLoader might then be a consumer of this.

**Comment by OBannon37** [comment]
I agree - I think DataLoader would be a good location to implement this. I think to narrow the scope, we could tune num_workers at first. This commit to TensorFlow might be a good place to start - https://github.com/tensorflow/tensorflow/commit/c8a0dfc741736a59f8fd1776b71f38619d66da56#diff-df634c8243713c0afd2e05c1689412e2

**Comment by apaszke** [comment]
Automatic tuning might be a bit hard in PyTorch, because we're generally trying to not make any assumptions about the user code, and our objects are very loosely decoupled (as opposed to TensorFlow, which has everything encoded in the graph). In any case, I don't think that tuning the batch size even makes sense, because changing it might have significant implications for the speed of convergence and quality of the resulting model.

**Comment by bhack** [comment]
Can we re-evaluate at least `num_workers` autotuning?
